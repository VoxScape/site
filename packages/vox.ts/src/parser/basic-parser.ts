import type * as VoxTypes from '../types/vox-types';
import { RiffLense } from '../util/riff-lense';
import { getDefaultPalette, readMatt, readRgba, readXyzi, uint32ToColor } from './chunk-reader';

/**
 * parser for "basic" vox format
 */
export function basicParser(
  bytes: ArrayBuffer,
  /** @deprecated prefer to swap yz in mesh builder */ flipYZ = false,
  enableExtension = false,
): VoxTypes.ParsedVoxFile {
  const l = new RiffLense(bytes);

  if (l.asciiAt(0, 4) !== 'VOX ') throw new Error(`expected 'VOX ' at offset=0`);
  if (l.int32At(4) !== 150) throw new Error(`expect 0x150 at offset=4`);

  const mainChunk = l.chunkAt(8);
  if (mainChunk.id !== 'MAIN') throw new Error('expected MAIN chunk');

  const mainChildren = l.childrenChunksIn(mainChunk);

  const models: VoxTypes.VoxelModel[] = [];
  let palette: undefined | VoxTypes.VoxelPalette = undefined;
  const materials: VoxTypes.VoxelMaterial[] = [];
  const warnings: string[] = [];
  for (let chunkIndex = 0; chunkIndex < mainChildren.length; ++chunkIndex) {
    const chunk = mainChildren[chunkIndex];
    if (chunk.id === 'PACK' && chunkIndex === 0) {
      // ignore
    } else if (chunk.id === 'SIZE') {
      const xyzi = mainChildren[chunkIndex + 1];
      if (!(xyzi && xyzi.id === 'XYZI')) {
        throw new Error(`expected 'XYZI' chunk at offset=0x${chunk.end.toString(16)}`);
      }
      models.push(readXyzi(l, chunk, xyzi, flipYZ));
      ++chunkIndex;
    } else if (chunk.id === 'RGBA') {
      palette = readRgba(l, chunk);
    } else if (chunk.id === 'MATT') {
      materials.push(readMatt(l, chunk));
    } else {
      switch (chunk.id) {
        case 'MATL':
        case 'nTRN':
        case 'nGRP':
        case 'nSHP':
        case 'LAYR':
        case 'rOBJ':
        case 'STRING':
        case 'DICT':
        case 'ROTATION':
        case 'NOTE':
        case 'IMAP': {
          const msg = `vox.ts: unsupported ${chunk.id} chunk at offset 0x${chunk.start.toString(16)}`;
          warnings.push(msg);
          break;
        }

        default: {
          warnings.push(`vox.ts: unknown ${chunk.id} chunk at offset 0x${chunk.start.toString(16)}`);
          break;
        }
      }
    }
  }

  return {
    models,
    palette: palette ?? getDefaultPalette(),
    materials,
    warnings,
    enableExtension,
  };
}
