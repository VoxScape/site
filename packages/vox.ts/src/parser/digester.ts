import { ParsedVoxFile, VoxelModel, VoxelModelSize } from '../types/vox-types';

export interface VoxFileDigest {
  path: string;
  warnings: readonly string[];
  models: {
    size: VoxelModelSize;
    numVoxels: number;
  }[];
}

export function digestFile(path: string, f: ParsedVoxFile): VoxFileDigest {
  return {
    path,
    warnings: f.warnings,
    models: f.models.map((m) => ({
      size: m.size,
      numVoxels: m.voxels.length,
    })),
  };
}
