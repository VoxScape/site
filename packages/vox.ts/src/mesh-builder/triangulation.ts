import type { Scene } from '@babylonjs/core/scene';
import { Mesh, VertexData } from '@babylonjs/core/Meshes';
// @ts-ignore
import zeros from 'zeros';
import ndarray from 'ndarray';
// @ts-ignore
import triangulateVoxels from 'voxel-triangulation';
import type * as Vox from '../types/vox-types';
import { getDefaultPalette } from '../parser/chunk-reader';
const debug = require('debug')('vox:mesh-builder:triangulation');

/**
 * mostly a port from https://github.com/FlorianFe/vox-viewer/blob/master/vox-viewer.js
 */
export async function buildTriangulatedMesh(f: Vox.VoxelModel, palette: Vox.VoxelPalette, scene: Scene, options?: {name: string}): Promise<Mesh> {
  const xyzc: ndarray.NdArray = zeros(f.size.x, f.size.y, f.size.z);
  f.voxels.forEach((v) => {
    xyzc.set(v.x, v.y, v.z, v.colorIndex);
  });
  /**
   * transpose to (y,z,x)=>color ndarray?
   * @xxx why is this required?
   * @see https://numpy.org/doc/stable/reference/generated/numpy.ndarray.transpose.html
   */
  const transposed = xyzc.transpose(1, 2, 0);

  /**
   * @see https://github.com/FlorianFe/voxel-triangulation.js
   */
  const triangulaized: {
    verices: number[];
    normals: number[];
    indices: number[];
    voxelValues: number[];
  } = triangulateVoxels(transposed);

  debug('triangulaized', triangulaized);

  const colors = expandColor(palette);

  /**
   * @see https://doc.babylonjs.com/features/featuresDeepDive/mesh/creation/custom/custom
   */
  const mesh = new Mesh(options?.name ?? 'voxel', scene);

  const vertexData = new VertexData();
  vertexData.positions = triangulaized.verices;
  vertexData.indices = triangulaized.indices;
  vertexData.normals = triangulaized.normals;
  vertexData.colors = colors;

  vertexData.applyToMesh(mesh);
  return mesh;
}

function expandColor(palette: Vox.VoxelPalette): Float32Array {
  const BYTE_PER_COLOR = 4; // RGBA
  const flatten = new Array<number>(palette.length * BYTE_PER_COLOR);

  for (let i = 0; i < palette.length; i++) {
    flatten[BYTE_PER_COLOR * i] = palette[i].r;
    flatten[BYTE_PER_COLOR * i + 1] = palette[i].g;
    flatten[BYTE_PER_COLOR * i + 2] = palette[i].b;
    flatten[BYTE_PER_COLOR * i + 3] = palette[i].a;
  }
  return new Float32Array(flatten);
}
