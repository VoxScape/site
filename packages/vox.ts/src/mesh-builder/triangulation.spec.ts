import { triangulate } from './triangulation';
import type * as Vox from '../types/vox-types';
import { getDefaultPalette } from '../parser/chunk-reader';
import { VertexData } from '@babylonjs/core/Meshes';
// @ts-ignore
import triangulateVoxels from 'voxel-triangulation';
// @ts-ignore
import zeros from 'zeros';
import ndarray from 'ndarray';

const demoModel: Vox.VoxelModel = {
  size: { x: 2, y: 2, z: 2 },
  voxels: [{ x: 1, y: 1, z: 1, colorIndex: 3 }],
};

function buildNdArray(f: Vox.VoxelModel): ndarray.NdArray {
  const xyzc: ndarray.NdArray = zeros([f.size.x, f.size.y, f.size.z], 'float');
  f.voxels.forEach((v) => {
    xyzc.set(v.x, v.y, v.z, v.colorIndex);
  });
  return xyzc.transpose(1, 2, 0);
}

describe.skip('ndarray', () => {
  it('builds an ndarray from a voxel model', () => {
    const xyzc = buildNdArray(demoModel);
    expect(xyzc).toMatchSnapshot('xyzc');
  });
});

describe.skip('triangulation', () => {
  it('triangulate voxel model into coordinates', () => {
    const triangulated = triangulate(demoModel, getDefaultPalette());

    expect(triangulated).toBeInstanceOf(VertexData);

    expect(triangulated.positions).toEqual([]);
  });
});
