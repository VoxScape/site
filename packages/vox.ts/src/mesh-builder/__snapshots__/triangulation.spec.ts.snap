// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildNdArray builds an ndarray from a voxel model: xyzc 1`] = `
View3dfloat32 {
  "data": Float32Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    3,
  ],
  "offset": 0,
  "shape": [
    2,
    2,
    2,
  ],
  "stride": [
    2,
    1,
    4,
  ],
}
`;

exports[`ndarray builds an ndarray from a voxel model: xyzc 1`] = `
View3dfloat32 {
  "data": Float32Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    3,
  ],
  "offset": 0,
  "shape": [
    2,
    2,
    2,
  ],
  "stride": [
    2,
    1,
    4,
  ],
}
`;
